src/runtime/HalideRuntime.h

```diff
@@ -439,6 +439,12 @@ struct halide_type_t {
         return !(*this == other);
     }
 
+    HALIDE_ALWAYS_INLINE bool operator<(const halide_type_t &other) const {
```

Comment:
There's a lot of care taken above to make sure a halide_type_t occupies 32 bits. I think the following is not UB and compiles to a single integer comparison op:
```
uint32_t as_u32() const {
  uint32_t u;
  memcpy(&u, this, sizeof(u));
  return u;
}

bool operator<(const halide_type_t &other) const {
  return as_u32() < other.as_u32();
}

bool operator==(const halide_type_t &other) const {
  return as_u32() == other.as_u32();
}
```

It's a modest increase in complexity/confusingness for a modest performance advantage, so I don't feel strongly about this suggestion.